### PR TITLE
GH-96793: Implement PEP 479 in bytecode.

### PR DESCRIPTION
**Pull request by markshannon**
In order to specialize iteration over generators and send to coroutines we need to remove the pre- and post checks from the C wrappers.
The pre-checks are largely eliminated by the the compiler and specialization.
The post-checks need to be handled in bytecode.
To do that, we wrap the generator body in a `try: ... except StopIteration: ...` which converts the `StopIteration` into a `RuntimeError`.

I initially did this entirely in bytecode, but it adds a [lot of bulk to the bytecode](https://github.com/python/cpython/commit/9cda4fcd71cb5967a34886c997a78072016611a5#diff-fd7b118a01a9695dfa16eea9381d3c7b0652e1b7b55f51ff052a1f620d175642R715).

The new `STOPITERATION_ERROR` instruction should only be temporary: https://github.com/python/cpython/issues/99005

* Issue: gh-96793

### CODE REVIEW

File: Lib/test/test_sys_settrace.py [hunk 1]
```diff
@@ -346,7 +346,7 @@ def make_tracer():
         return Tracer()
 
     def compare_events(self, line_offset, events, expected_events):
-        events = [(l - line_offset, e) for (l, e) in events]
+        events = [(l - line_offset if l is not None else None, e) for (l, e) in events]
```
Comment (markshannon) [inline Lib/test/test_sys_settrace.py:349]: I introduced for debugging this PR. It is no longer strictly necessary, but it gives nicer output when tests fail.

File: Python/compile.c [hunk 2]
```diff
@@ -2567,6 +2570,42 @@ compiler_check_debug_args(struct compiler *c, arguments_ty args)
...
+    for (int i = block->b_iused - 1; i > pos; i--) {
+        block->b_instr[i] = block->b_instr[i-1];
+    }
+    block->b_instr[pos] = *instr;
+    return 0;
+}
+
+static int
+wrap_in_stopiteration_handler(struct compiler *c)
+{
```
Comment (iritkatriel) [inline Python/compile.c:2565]: I wonder if this is not wrapping too much, like the instructions that deal with the function's arguments and annotations?

Comment (markshannon) [reply Python/compile.c:2565]: Annotations are executed in the enclosing scope, as are defaults. So there is no problem there.
`MAKE_CELL` and `RETURN_GENERATOR` instructions are inserted in the back-end so happen after this.

**Review by iritkatriel — APPROVED** [review]
LGTM

**Pull request merged**